GSuite integration

almeidapaulopt opened the pull request:
This integration gives the user a possibility to create GSuite Documents with data from a frappe doctype.

First deploy the script frappe/integrations/doctype/gsuite_settings/gsuite_settings.gs as web app on google apps script, get the URL.

The frappe configurations is on config/integrations
![gsuite1](https://cloud.githubusercontent.com/assets/22657407/25847084/8422b7fa-34ac-11e7-9649-d4da70fe3220.png)

Configure Gsuite with URL of you gsuite web app and the credentials from google
![gsuite2](https://cloud.githubusercontent.com/assets/22657407/25847099/9bfeca4e-34ac-11e7-8481-60eaf41d6343.png)
Test if the script is correctly installed, after "alow Gsuite access".

the configure the templates for each doctype.
![gsuite3](https://cloud.githubusercontent.com/assets/22657407/25847175/e0de3582-34ac-11e7-8461-0274237e0558.png)

create a document in google docs, Use de google file_id as template, the name of the file can use fields from doctype, and destination is the google drive folder where the file will be created (if empty saves on root)
In the document use {{FIELD}} with FIELD as field of doctype.
![gsuite4](https://cloud.githubusercontent.com/assets/22657407/25847248/36cfde32-34ad-11e7-8215-65ee87000b34.png)

then on the frappe document "create GSuite document" ,select the template and create the file.
![gsuite_upload](https://cloud.githubusercontent.com/assets/22657407/25958329/40ca85ce-3668-11e7-86fb-05f52ba2ec64.png)


The file is saved as a attachment of the document and the comments are created too.

![gsuite6](https://cloud.githubusercontent.com/assets/22657407/25847441/19df5158-34ae-11e7-9cce-69c4953cbb21.png)




**NOTES**
Only google documents is supported. sheets and slides TBD.
Delete the attachment doesn't delete the file don google drive (should?)
Manual/Documentation TDB

almeidapaulopt committed:
GSuite initial commit

almeidapaulopt committed:
GSuite initial commit

almeidapaulopt committed:
cleanups

almeidapaulopt committed:
exception handle

almeidapaulopt committed:
gsuite script: add doGet

rmehta commented:
@almeidapaulopt looks nice. See my comments

rmehta commented on frappe/public/js/frappe/form/footer/attachments.js:
```diff
@@ -13,6 +13,9 @@ frappe.ui.form.Attachments = Class.extend({
 		this.parent.find(".add-attachment").click(function() {
 			me.new_attachment();
 		});
+		this.parent.find(".add-gsuite").click(function() {
```
This should be in the attachment popup, we can directly have an option to upload to GSuite or any other service in a generic way. Can you do that?

almeidapaulopt replied:
As this isn't an upload (the template is already in google drive), we should change the popup title "upload attachment" to something like "Add Attachment" to be more service agnostic.
It seems to me that this is a major change js files (as far i found, the click on attach button is defined in frappe.upload.make. Do you think I should only add there the code for Gsuite or you what to make any kind of abstraction?

almeidapaulopt replied:
removed and already integrated in popup

rmehta commented on frappe/public/js/frappe/form/templates/form_sidebar.html:
```diff
@@ -44,7 +44,9 @@
 	<li><a class="strong add-attachment">{%= __("Attach File") %}
-		<i class="octicon octicon-plus" style="margin-left: 2px;"></i></li></a>
+		<i class="octicon octicon-plus" style="margin-left: 2px;"></i></a></li>
+	<li><a class="strong add-gsuite">{%= __("Create GSuite Document") %}
```
this cannot be a permanent fixture, see my earlier comment

almeidapaulopt replied:
removed

rmehta reviewed (changes requested):
Add the docs!

almeidapaulopt committed:
Add GSuite integration to the upload popup

rmehta commented on frappe/public/js/frappe/upload.js:
```diff
@@ -374,5 +375,17 @@ frappe.upload = {
 			d.show();
 			opts.confirm_is_private =  0;
+	},
+	create_gsuite_file: function(args, opts) {
```
lets put this function in the gsuite js file, since its an external integration

rmehta commented on frappe/public/js/frappe/form/footer/attachments.js:
```diff
@@ -209,11 +209,13 @@ frappe.ui.form.Attachments = Class.extend({
+			{"fieldtype": "Section Break"},
 			{"fieldtype": "Link" , "fieldname": "file" , "label": __("Select uploaded file"), "options": "File"},
-	    ],
+			{"fieldtype": "Section Break", "label": __("Gsuite Document")},
```
these fields should be hidden by default unless gsuite is integrated. You can add a property in `boot.py` that you can access as `frappe.boot`

rmehta commented on frappe/integrations/doctype/gsuite_settings/gsuite_settings.gs:
```diff
@@ -0,0 +1,49 @@
+function doGet(e){
```
why is this extension `gs` ?

almeidapaulopt replied:
this is what will be installed in google apps script (https://script.google.com/)

rmehta reviewed (changes requested):
Great. thanks for revert. Just one more thing, the gsuite section should be hidden if gsuite is not activated.

almeidapaulopt committed:
hide dialog fields if gsuite disabled

almeidapaulopt committed:
move gsuite code to gsuite.js from upload.js

almeidapaulopt committed:
documentation

almeidapaulopt committed:
change name format

rmehta commented on frappe/docs/user/en/guides/integration/google_gsuite.md:
```diff
@@ -0,0 +1,68 @@
+
+1. Go to [https://script.google.com](https://script.google.com)
+1. Create a new Project. Click on **File > New > Project**
+1. Fill Code.gs the content of ***<bench dir>/apps/frappe/integrations/doctype/gsuite_settings/gsuite_settings.gs***
```
I am sure there must be a better way to do this! I don't think 99% of the users will be able to figure this out.

almeidapaulopt replied:
@rmehta 
you have 2 other options:
1. Publish a webapp a define as default in frappe (but google set limits that are very high for a single company but with a lot of user it could be a problem)
1. Use as a library (that google don't recommend because performance issues), but it is not that simplified that justify a change for a copy-paste.

please be sure I follow your concerns, but:
1. I've tried to install the webapp with REST, but google don't give chance to install, just run installed apps
1. Do you really think that an administrator of an ERP or a developer with frappe framework will not be able to follow a simple script of steps with a couple of copy paste? This isn't that different that activating google Social Login (create a project, copy credentials and past in frappe)...this integration just need to make a copy paste of a file before.

rmehta replied:
Most erpnext users don't understand the need between a `git stash` and `git commit`. Asking users to change a file is a complete no no. For all integrations all you need is to update your API keys.

Can't you just include that file as a template? And replace the dynamic part?

almeidapaulopt replied:
I'm not asking to change any file, maybe I should move te .gs file to the documentation. because what I'm asking is to copy the content of .gs file to script.google.com. .gs file is the template in this case.
Of course we couldn't ask a user to change any file on frappe.

almeidapaulopt replied:
@rmehta 
I've changed the docs to be more comprehensive (I hope :) )

rmehta commented on frappe/public/js/frappe/form/footer/attachments.js:
```diff
@@ -209,11 +209,13 @@ frappe.ui.form.Attachments = Class.extend({
+			{"fieldtype": "Section Break"},
 			{"fieldtype": "Link" , "fieldname": "file" , "label": __("Select uploaded file"), "options": "File"},
-	    ],
+			{"hidden": !frappe.boot.gsuite_enabled ,"fieldtype": "Section Break", "label": __("Gsuite Document")},
```
GSuite (s is capital)

almeidapaulopt committed:
typo on GSuite

almeidapaulopt committed:
copy paste the file .gs

almeidapaulopt committed:
fix:don't show gsuite when creating a file in filemanager

almeidapaulopt committed:
add version to webapp

rmehta commented:
So I am not sure what to do here. I don't think any integration that allows editing of documents should be allowed. Also this is a very specific gsuite implementation. 

On the other hand, I would he happy to allow this if this is a start and maybe others can build on this and make it simpler to use. 

Let me sleepover this and keep it open for another day to decide.

felixvarghese commented:
I am also not a fan of having the user create a file in this way.

Maybe create a longtext field in a Settings page and have the user paste the script in that, and have things work off that. That seems to be a middle ground.

almeidapaulopt commented:
@felixvarghese It isn't asked to user create any file. I think that is obvious in documentation I've sent, please inform me if it isn't. I'll change it to be more understandable.

almeidapaulopt commented:
@felixvarghese and btw, you paste the script in Google apps site not in frappe.

@rmehta the need of a Google apps script is because Google only publish google documents API to that. Sheets for example is public and you can use REST. But I'll continue to find other options (even after you accept this).

felixvarghese commented:
Ahh, yes, I understand now. Maybe use the following

Copy the content of ***<bench dir>/apps/frappe/integrations/doctype/gsuite_settings/gsuite_settings.gs*** into Code.gs in your new Google Project.

almeidapaulopt committed:
add webapp default url

almeidapaulopt commented:
@rmehta I've added a publish google webapp and set as default and updated the docs. Right now the user can use both a own publish google webapp or the default and to that user doesn't need to do nothing.
What do think about this?

rmehta commented:
@almeidapaulopt ok now i understood, its the other way round!. Here is what I would suggest, paste the contents of the file in a `<code>` block in an HTML field in **GSuite Settings** and ask the user to copy paste it from there! Also add the setup steps there itself.

That will make it much easier for users!

almeidapaulopt commented:
@rmehta Good Idea.

what about this?
![gs1](https://cloud.githubusercontent.com/assets/22657407/26251288/b3f388e2-3ca4-11e7-9e29-48d1a5f31cca.png)


![gs2](https://cloud.githubusercontent.com/assets/22657407/26251297/b9074968-3ca4-11e7-8fae-15f7abfdad02.png)

almeidapaulopt committed:
move webapp to html fied

almeidapaulopt committed:
fix rebase

almeidapaulopt committed:
fixes for codecy

rmehta commented:
@almeidapaulopt should I merge this?

almeidapaulopt commented:
@rmehta yes I think so.

rmehta commented:
@almeidapaulopt done. From now on please send smaller PRs! 

https://github.com/frappe/erpnext/wiki/Cascading-Pull-Requests
